libs/pubsub/query/syntax/parser.go

```diff
@@ -0,0 +1,213 @@
+package syntax
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Parse parses the specified query string. It is shorthand for constructing a
+// parser for s and calling its Parse method.
+func Parse(s string) (Query, error) {
+	return NewParser(strings.NewReader(s)).Parse()
+}
+
+// Query is the root of the parse tree for a query.  A query is the conjunction
+// of one or more conditions.
+type Query []Condition
+
+func (q Query) String() string {
+	ss := make([]string, len(q))
+	for i, cond := range q {
+		ss[i] = cond.String()
+	}
+	return strings.Join(ss, " AND ")
+}
+
+// A Condition is a single conditional expression, consisting of a tag, a
+// comparison operator, and an optional argument. The type of the argument
+// depends on the operator.
+type Condition struct {
+	Tag string
+	Op  Token
+	Arg *Arg
+
+	opText string
+}
+
+func (c Condition) String() string {
+	s := c.Tag + " " + c.opText
+	if c.Arg != nil {
+		return s + " " + c.Arg.String()
+	}
+	return s
+}
+
+// An Arg is the argument of a comparison operator.
+type Arg struct {
+	Type Token
+	text string
+}
+
+func (a *Arg) String() string {
+	if a == nil {
+		return ""
+	}
+	switch a.Type {
+	case TString:
+		return "'" + a.text + "'"
+	case TTime:
+		return "TIME " + a.text
+	case TDate:
+		return "DATE " + a.text
+	default:
+		return a.text
+	}
+}
+
+// Number returns the value of the argument text as a number, or a NaN if the
+// text does not encode a valid number value.
+func (a *Arg) Number() float64 {
+	if a == nil {
+		return -1
+	}
+	v, err := strconv.ParseFloat(a.text, 64)
+	if err == nil && v >= 0 {
+		return v
+	}
+	return math.NaN()
+}
+
+// Time returns the value of the argument text as a time, or the zero value if
+// the text does not encode a timestamp or datestamp.
+func (a *Arg) Time() time.Time {
```

Comment:
pointer so that we can avoid the need for time.IsZero checks

Reply:
> This doesn't change any query parsing behavior right?

That's correct. Or at least, that's the intent, and the tests from the original implementation agree except in the one case I explicitly chose not to support.

Reply:
Re: Pointer vs. value. If we make it a pointer, the caller basically _always_ has to check. The point here is to have a graceful fallback if the caller happens to call it on an arbitrary token. In common use, you don't call this method unless you know the argument already has a Time, in which case you never see the fallback.

I'd rather not convert that into a panic, but I could see an argument the other way.